### DISCUSSION TITLE
End 2 End streaming decryption does NOT work with node.js?

### DISCUSSION THREAD

**Discussion by fbx31** [openpgpjs/openpgpjs#1360] [Q&A] [answered]
Hello,
Since several days now, I am trying to implement a full end to end decryption stream using v5 of openpgpjs.
RSA 2048 uncompressed algorithm used, key pairs generated with 5.0 and encryption/decryption working fine.
BUT when i create a stream to read encrypted file, calling decrypt passing the message readable stream and piping the result.data to file writable stream, this is what happens:
1. The first file is read entirely in memory (packet by packet)
2.  The decryption is launched ONCE the file is entirely loaded and starts emitting packet by packet to the writable file.
Second phase is fine but the first one let me think that the decryption does not support streaming for decryption on the fly.

Am i wrong? Is it related to the encryption algorithm used? A bug  in the v5? A bug in my code ? 

Here is the code (inside a promise):

                let file = fs.createReadStream(record.input, { encoding: null, flags: "r" });
                file.on("data", (d) => console.log("Reading...", d.length));
                file.on("end", () => { console.log("End read"); resolve(); });
                file.on("error", (err) => reject(err));

                const message = await openpgp.readMessage({ binaryMessage: file });
                const decrypt = await openpgp.decrypt({
                    message,,
                    format: "binary",
                    decryptionKeys: privateKey,
                    config: { allowInsecureDecryptionWithSigningKeys: true }
                });
                let decryptStream = decrypt.data;
                console.log("end decrypt");

                decryptStream.on("data", (d) => console.log("Decrypting...", d.length));
                decryptStream.on("end", () => { console.log("End decrypt"); resolve(); });
                decryptStream.on("error", (err) => reject(err));

                let writeStream = fs.createWriteStream(record.output, { encoding: null, flags: "w" });
                writeStream.on("drain", () => console.log("Writing..."));
                writeStream.on("finish", () => console.log("End writing"));

                decryptStream.pipe(writeStream);

Thanks a lot for any help as, for now, I am stucked....

**Comment by fbx31** [comment, marked as answer]
OK just find the solution, adding allowUnauthenticatedStream: true in configuration solved the issue.
Quite strange and bad documented by the way.

**Reply by larabr** [reply]
Hi @fbx31, by default, the decrypted data is not released until everything is fully decrypted. This is for security reasons, since we can only confirm that the message was not corrupted after decrypting everything.
`config.allowUnauthenticatedStream` releases each chunk as soon as it is decrypted, hence it does help reduce memory usage, but we don't really recommend enabling it in the general case. As long as you are just writing the data to a file, and not processing the partially decrypted data further (for example, by rendering a partially decrypted HTML email), then it should be fine.

About documentation, we will probably create a wiki entry about streaming, since we understand that the library behaviour can be a bit confusing :)

**Comment by fbx31** [comment]
Thanks for the explanation and the action taken to clarify the documentation. Will save time of a lot of people like me.
Congrats also for the job done in v5.
May I suggest to add a PGP for newbie in docs explaining to people what are publid, private, session keys, signatures and verifications... Should avoid also bad usage.
Thankd
